# u2001820 Powerlifting Project: Data Cleaning
---
We aim to investigate the powerlifting dataset available on <a href="https://www.kaggle.com/datasets/open-powerlifting/powerlifting-database">Kaggle</a>.

In [1]:
# imports
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv("openpowerlifting.csv")

C:\Users\asale\AppData\Local\Temp\ipykernel_12156\2594218218.py:1: DtypeWarning: Columns (35) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("openpowerlifting.csv")


In [3]:
df.head()

Name Sex Event Equipment   Age AgeClass Division  BodyweightKg  \
0     Abbie Murphy   F   SBD     Wraps  29.0    24-34     F-OR          59.8   
1      Abbie Tuong   F   SBD     Wraps  29.0    24-34     F-OR          58.5   
2   Ainslee Hooper   F     B       Raw  40.0    40-44     F-OR          55.4   
3  Amy Moldenhauer   F   SBD     Wraps  23.0    20-23     F-OR          60.0   
4     Andrea Rowan   F   SBD     Wraps  45.0    45-49     F-OR         104.0   

  WeightClassKg  Squat1Kg  ...  McCulloch  Glossbrenner  IPFPoints  Tested  \
0            60      80.0  ...     324.16        286.42     511.15     NaN   
1            60     100.0  ...     378.07        334.16     595.65     NaN   
2            56       NaN  ...      38.56         34.12     313.97     NaN   
3            60    -105.0  ...     345.61        305.37     547.04     NaN   
4           110     120.0  ...     338.91        274.56     550.08     NaN   

   Country  Federation        Date  MeetCountry  MeetState       MeetName  
0      NaN     GPC-AUS  2018-10-27    Australia        VIC  Melbourne Cup  
1      NaN     GPC-AUS  2018-10-27    Australia        VIC  Melbourne Cup  
2      NaN     GPC-AUS  2018-10-27    Australia        VIC  Melbourne Cup  
3      NaN     GPC-AUS  2018-10-27    Australia        VIC  Melbourne Cup  
4      NaN     GPC-AUS  2018-10-27    Australia        VIC  Melbourne Cup  

[5 rows x 37 columns]

In [6]:
df.describe()

Age  BodyweightKg       Squat1Kg       Squat2Kg  \
count  757527.000000  1.406622e+06  337580.000000  333349.000000   
mean       31.501570  8.422503e+01     114.102442      92.155846   
std        13.371707  2.322011e+01     147.143021     173.701524   
min         0.000000  1.510000e+01    -555.000000    -580.000000   
25%        21.000000  6.670000e+01      90.000000      68.000000   
50%        28.000000  8.180000e+01     147.500000     145.000000   
75%        40.000000  9.915000e+01     200.000000     205.000000   
max        97.000000  2.580000e+02     555.000000     566.990000   

            Squat3Kg     Squat4Kg  Best3SquatKg       Bench1Kg       Bench2Kg  \
count  323842.000000  3696.000000  1.031450e+06  499779.000000  493486.000000   
mean       30.056842    71.356870  1.740049e+02      83.892373      55.065745   
std       200.413385   194.522045  6.923931e+01     105.196350     130.302229   
min      -600.500000  -550.000000 -4.775000e+02    -480.000000    -507.500000   
25%      -167.500000  -107.840000  1.224700e+02      57.500000     -52.500000   
50%       110.000000   135.000000  1.678300e+02     105.000000      95.000000   
75%       192.500000   205.000000  2.175000e+02     145.000000     145.000000   
max       560.000000   505.500000  5.750000e+02     467.500000     487.500000   

            Bench3Kg  ...    Deadlift1Kg    Deadlift2Kg    Deadlift3Kg  \
count  478485.000000  ...  363544.000000  356023.000000  339947.000000   
mean      -18.520481  ...     162.700840     130.228378      12.995484   
std       144.225726  ...     108.681438     162.680134     215.052488   
min      -575.000000  ...    -461.000000    -470.000000    -587.500000   
25%      -140.000000  ...     125.000000     115.000000    -210.000000   
50%       -60.000000  ...     180.000000     177.500000     117.500000   
75%       117.500000  ...     226.800000     230.000000     205.000000   
max       478.540000  ...     450.000000     460.400000     457.500000   

       Deadlift4Kg  Best3DeadliftKg       TotalKg         Wilks     McCulloch  \
count  9246.000000     1.081808e+06  1.313184e+06  1.304407e+06  1.304254e+06   
mean     78.914945     1.872585e+02  3.956148e+02  2.882247e+02  2.960682e+02   
std     192.605159     6.232821e+01  2.011420e+02  1.231805e+02  1.249700e+02   
min    -461.000000    -4.100000e+02  2.500000e+00  1.470000e+00  1.470000e+00   
25%    -110.000000     1.383500e+02  2.325000e+02  1.979000e+02  2.048200e+02   
50%     145.150000     1.850000e+02  3.787500e+02  3.052000e+02  3.120300e+02   
75%     210.000000     2.300000e+02  5.400000e+02  3.745600e+02  3.837600e+02   
max     418.000000     5.850000e+02  1.367500e+03  7.793800e+02  8.044000e+02   

       Glossbrenner     IPFPoints  
count  1.304407e+06  1.273286e+06  
mean   2.718484e+02  4.854330e+02  
std    1.175571e+02  1.133489e+02  
min    1.410000e+00  2.160000e+00  
25%    1.828100e+02  4.028600e+02  
50%    2.859400e+02  4.780500e+02  
75%    3.552800e+02  5.597000e+02  
max    7.429600e+02  1.245930e+03  

[8 rows x 22 columns]

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1423354 entries, 0 to 1423353
Data columns (total 37 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   Name             1423354 non-null  object 
 1   Sex              1423354 non-null  object 
 2   Event            1423354 non-null  object 
 3   Equipment        1423354 non-null  object 
 4   Age              757527 non-null   float64
 5   AgeClass         786800 non-null   object 
 6   Division         1415176 non-null  object 
 7   BodyweightKg     1406622 non-null  float64
 8   WeightClassKg    1410042 non-null  object 
 9   Squat1Kg         337580 non-null   float64
 10  Squat2Kg         333349 non-null   float64
 11  Squat3Kg         323842 non-null   float64
 12  Squat4Kg         3696 non-null     float64
 13  Best3SquatKg     1031450 non-null  float64
 14  Bench1Kg         499779 non-null   float64
 15  Bench2Kg         493486 non-null   float64
 16  Bench3Kg         4

## Dropping rows/data imputation
The raw dataset contains 1,423,354 rows of data. The dataset is a combination of two smaller datasets. We now drop rows/impute values according to the following:
* Dropping Events other than 'SBD', as we are interested in individuals competing in all three major lifts.
* Dropping rows with both Age and AgeClass empty for simplicity. If a row has an age but not an age class, then we assign the correct AgeClass. In the converse case, we take the average age of the age class and assign that as the lifter's age.
* Removing the 4th attempt columns, because only the first 3 attempts of each candidate count towards total weight lifted and competition placement.
* Best squat, bench and deadlift: there are a lot of values in the best lifts columns that have not been filled in, which we take care of below, simply by taking the maximum of the competitors' 3 qualifying lifts in each area.
* Filling in the TotalKg: once the best squat, bench and deadlift values has been imputed/filled in for every competitor, we can fill in any missing TotalKg values simply by summing up the values in the three 'best' columns.

### Filtering Events

In [27]:
df_new = df[df["Event"] == "SBD"]

### Filtering Age and AgeClass

In [28]:
df_new['Age'].isnull().sum()

573802

In [29]:
df_new['Age'].value_counts()

Age
17.5    19044
22.5    17670
21.5    16994
18.5    16539
20.5    15923
        ...  
2.0         2
95.5        1
93.5        1
3.0         1
88.0        1
Name: count, Length: 186, dtype: int64

In [30]:
df_new['AgeClass'].unique()

array(['24-34', '20-23', '45-49', '35-39', '18-19', '50-54', '40-44',
       '16-17', '80-999', '70-74', nan, '65-69', '13-15', '60-64',
       '55-59', '75-79', '5-12'], dtype=object)

In [31]:
df_new['AgeClass'].isnull().sum()

555719

In [32]:
df_new['AgeClass'].value_counts()

AgeClass
24-34     165935
20-23     105165
18-19      50053
35-39      38695
40-44      35197
16-17      34928
45-49      24116
13-15      19372
50-54      18187
55-59      10487
60-64       6837
65-69       3803
70-74       2266
5-12        1344
75-79        873
80-999       260
Name: count, dtype: int64

In [33]:
# dropping rows where BOTH Age and AgeClass are empty
df_new = df_new[(df_new['Age'].isnull() == False) | (df_new['AgeClass'].isnull() == False)]

In [34]:
df_new['Age'].isnull().sum()

18185

In [35]:
df_new['AgeClass'].isnull().sum()

102

In [36]:
age_classes = df_new['AgeClass'].unique().tolist()
print(age_classes)

['24-34', '20-23', '45-49', '35-39', '18-19', '50-54', '40-44', '16-17', '80-999', '70-74', '65-69', '13-15', '60-64', '55-59', '75-79', '5-12', nan]


In [37]:
# taking the average age of each age class
ages = [29, 21.5, 47, 37, 18.5, 52, 42, 16.5, 80, 72, 67, 14, 62, 57, 77, 8.5]
classes = age_classes[:-1]

age_to_class = dict(zip(ages, classes))
class_to_age = dict(zip(classes, ages))
print(class_to_age)
print(age_to_class)

{'24-34': 29, '20-23': 21.5, '45-49': 47, '35-39': 37, '18-19': 18.5, '50-54': 52, '40-44': 42, '16-17': 16.5, '80-999': 80, '70-74': 72, '65-69': 67, '13-15': 14, '60-64': 62, '55-59': 57, '75-79': 77, '5-12': 8.5}
{29: '24-34', 21.5: '20-23', 47: '45-49', 37: '35-39', 18.5: '18-19', 52: '50-54', 42: '40-44', 16.5: '16-17', 80: '80-999', 72: '70-74', 67: '65-69', 14: '13-15', 62: '60-64', 57: '55-59', 77: '75-79', 8.5: '5-12'}


In [38]:
# Filling missing values in 'Age' based on 'AgeClass' using the dictionary
df_new['Age'] = df_new.apply(lambda row: class_to_age[row['AgeClass']] if pd.isnull(row['Age']) else row['Age'], axis=1)

In [39]:
df_new['Age'].isnull().sum()

0

In [40]:
# figuring out how to fill in the AgeClass given the age
target_value = 15
closest_key = min(age_to_class, key=lambda x: abs(x - target_value))

print(f"The key closest to {target_value} is: {closest_key}")

The key closest to 15 is: 14


In [41]:
# Filling missing values in 'AgeClass' based on 'Age' using smallest distance away from the key
df_new['AgeClass'] = df_new.apply(lambda row: age_to_class[min(age_to_class, key=lambda x: abs(x - row['Age']))] if pd.isnull(row['AgeClass']) else row['AgeClass'], axis=1)

In [42]:
df_new['AgeClass'].isnull().sum()

0

### Dropping the 4th attempts of SBD

In [43]:
print(df_new['Squat4Kg'].isnull().sum())
print(df_new['Bench4Kg'].isnull().sum())
print(df_new['Deadlift4Kg'].isnull().sum())

514423
514891
512913


In [44]:
columns_to_drop = ['Squat4Kg', 'Bench4Kg', 'Deadlift4Kg']
df_new = df_new.drop(columns = columns_to_drop)

### Filling in best squat, bench and deadlift

In [45]:
print(df_new['Best3SquatKg'].isnull().sum())
print(df_new['Best3BenchKg'].isnull().sum())
print(df_new['Best3DeadliftKg'].isnull().sum())

15025
19984
21417


In [46]:
df_new['Best3SquatKg'] = df_new.apply(lambda x: max(x['Squat1Kg'], x['Squat2Kg'], x['Squat3Kg']) if pd.isnull(x['Best3SquatKg']) else x['Best3SquatKg'], axis=1)

In [47]:
df_new[['Squat1Kg', 'Squat2Kg', 'Squat3Kg', 'Best3SquatKg']]

Squat1Kg  Squat2Kg  Squat3Kg  Best3SquatKg
0            80.0      92.5     105.0         105.0
1           100.0     110.0     120.0         120.0
3          -105.0    -105.0     105.0         105.0
4           120.0     130.0     140.0         140.0
5           127.5     135.0     142.5         142.5
...           ...       ...       ...           ...
1423349     160.0     170.0     175.0         175.0
1423350      90.0     100.0     110.0         110.0
1423351     140.0     150.0     160.0         160.0
1423352    -165.0    -165.0    -165.0        -165.0
1423353    -115.0     115.0    -120.0         115.0

[517620 rows x 4 columns]

In [48]:
df_new['Best3BenchKg'] = df_new.apply(lambda x: max(x['Bench1Kg'], x['Bench2Kg'], x['Bench3Kg']) if pd.isnull(x['Best3BenchKg']) else x['Best3BenchKg'], axis=1)
df_new['Best3DeadliftKg'] = df_new.apply(lambda x: max(x['Deadlift1Kg'], x['Deadlift2Kg'], x['Deadlift3Kg']) if pd.isnull(x['Best3DeadliftKg']) else x['Best3DeadliftKg'], axis=1)

We remove records in which the lifter failed all three attempts at the squat, bench or deadlift. In these cases, the competitor does not place well and so is of little interest to us.

In [49]:
df_new = df_new[(df_new['Best3SquatKg'] >= 0) & (df_new['Best3BenchKg'] >= 0) & (df_new['Best3DeadliftKg'] >= 0)]

In [50]:
# checking imputations
print(df_new['Best3SquatKg'].isnull().sum())
print(df_new['Best3BenchKg'].isnull().sum())
print(df_new['Best3DeadliftKg'].isnull().sum())

0
0
0


### Filling in an missing TotalKg values

In [51]:
df_new['TotalKg'].isnull().sum()

801

In [52]:
df_new['TotalKg'] = df_new['Best3SquatKg'] + df_new['Best3BenchKg'] + df_new['Best3DeadliftKg']

In [53]:
df_new['TotalKg'].isnull().sum()

0

In [54]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Index: 487537 entries, 0 to 1423351
Data columns (total 34 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   Name             487537 non-null  object 
 1   Sex              487537 non-null  object 
 2   Event            487537 non-null  object 
 3   Equipment        487537 non-null  object 
 4   Age              487537 non-null  float64
 5   AgeClass         487537 non-null  object 
 6   Division         484515 non-null  object 
 7   BodyweightKg     486038 non-null  float64
 8   WeightClassKg    486327 non-null  object 
 9   Squat1Kg         264615 non-null  float64
 10  Squat2Kg         261703 non-null  float64
 11  Squat3Kg         254730 non-null  float64
 12  Best3SquatKg     487537 non-null  float64
 13  Bench1Kg         264567 non-null  float64
 14  Bench2Kg         262197 non-null  float64
 15  Bench3Kg         254677 non-null  float64
 16  Best3BenchKg     487537 non-null  float64


## Extensions:
---
* Imputing rows where both Age and AgeClass are missing: could build a small ML model to impute the age value based on a subset of the dataset's features. For example, the Division and TotalKg columns could be used to infer the age of the competitor, as very old/very young participants will most likely be unable to exceed certain feats of strength compared to those aged between, say, 20 and 35.

### Extras

In [ ]:
# NO LONGER NEEDED
# def age_to_class_func(x):
#     if 24 <= x and x <= 34:
#         return age_classes[0]
#     elif 20 <= x and x <= 23:
#         return age_classes[1]
#     elif 45 <= x and x <= 49:
#         return age_classes[2]
#     elif 35 <= x and x <= 39:
#         return age_classes[3]
#     elif 18 <= x and x <= 19:
#         return age_classes[4]
#     elif 50 <= x and x <= 54:
#         return age_classes[5]
#     elif 40 <= x and x <= 44:
#         return age_classes[6]
#     elif 16 <= x and x <= 17:
#         return age_classes[7]
#     elif 70 <= x and x <= 74:
#         return age_classes[9]
#     elif 65 <= x and x <= 69:
#         return age_classes[10]
#     elif 13 <= x and x <= 15:
#         return age_classes[11]
#     elif 60 <= x and x <= 64:
#         return age_classes[12]
#     elif 55 <= x and x <= 59:
#         return age_classes[13]
#     elif 75 <= x and x <= 79:
#         return age_classes[14]
#     elif 5 <= x and x <= 12:
#         return age_classes[15]
#     else:
#         return age_classes[8]